### PR TITLE
Adds remove_approval feature to treasury pallet

### PR DESCRIPTION
**Pull request by guzzit**
This PR adds a feature to remove previously approved proposals from the approval queue.

Resolves [Force Remove Approved Extrinsic in Treasury #10952](https://github.com/paritytech/substrate/issues/10952)

polkadot address: 13dUB1KQNcbCGAxbEpzFBsnbvkwVAjmHtBj41E6oPfmfpaAX

polkadot companion: paritytech/polkadot#5208[](https://github.com/paritytech/polkadot/pull/5208)

### CODE REVIEW

File: frame/treasury/src/lib.rs [hunk 1]
```diff
@@ -393,6 +395,38 @@ pub mod pallet {
...
+				if let Some(index) = v.iter().position(|x| x == &proposal_id) {
+					v.remove(index);
+					Ok(())
+				}
+				else{
+					Err(())
+				}
+			});
+
+			ensure!(proprosal_removed.is_ok(), Error::<T, I>::ProposalNotApproved);
```
Comment (xlc) [inline frame/treasury/src/lib.rs:426]:
```suggestion
			Approvals::<T, I>::try_mutate(|v| -> DispatchResult {
				if let Some(index) = v.iter().position(|x| x == &proposal_id) {
					v.remove(index);
					Ok(())
				}
				else{
					Err(Error::<T, I>::ProposalNotApproved.into())
				}
			})?;
```

File: frame/treasury/src/lib.rs [hunk 2]
```diff
@@ -393,6 +395,36 @@ pub mod pallet {
...
+		/// - DbWrites: `Approvals`
+		/// # </weight>
+		#[pallet::weight((T::WeightInfo::remove_approval(), DispatchClass::Operational))]
+		pub fn remove_approval(
+			origin: OriginFor<T>,
+			#[pallet::compact] proposal_id: ProposalIndex,
+		) -> DispatchResult {
+			T::RejectOrigin::ensure_origin(origin)?;
+
+			ensure!(<Proposals<T, I>>::contains_key(proposal_id), Error::<T, I>::InvalidIndex);
```
Comment (shawntabrizi) [inline frame/treasury/src/lib.rs:415]: Doesn't seem to me like we really need this check?

If it exists in the `Approvals`, that is all we want to do to remove it.

File: frame/treasury/src/lib.rs [hunk 3]
```diff
@@ -393,6 +395,36 @@ pub mod pallet {
...
+			Approvals::<T, I>::try_mutate(|v| -> DispatchResult {
+				if let Some(index) = v.iter().position(|x| x == &proposal_id) {
+					v.remove(index);
```
Comment (shawntabrizi) [inline frame/treasury/src/lib.rs:418]: We also need to `take` on `Proposals` storage here.

File: frame/treasury/src/tests.rs [hunk 4]
```diff
@@ -388,3 +388,15 @@ fn max_approvals_limited() {
...
+fn remove_already_removed_approval_fails() {
+	new_test_ext().execute_with(|| {
+		Balances::make_free_balance_be(&Treasury::account_id(), 101);
+
+		assert_ok!(Treasury::propose_spend(Origin::signed(0), 100, 3));
+		assert_ok!(Treasury::approve_proposal(Origin::root(), 0));
+		assert_ok!(Treasury::remove_approval(Origin::root(), 0));
+		assert_noop!(Treasury::remove_approval(Origin::root(), 0), Error::<Test, _>::ProposalNotApproved);
+	});
+}
```
Comment (shawntabrizi) [inline frame/treasury/src/tests.rs:402]: trailing newline please

**Review by shawntabrizi — APPROVED** [review]
please include your polkadot/kusama account information in the description to get a tip.

File: frame/treasury/src/lib.rs (hunk 2 above)
Comment (guzzit) [reply frame/treasury/src/lib.rs]: Resolved

File: frame/treasury/src/tests.rs (hunk 4 above)
Comment (guzzit) [reply frame/treasury/src/tests.rs]: Resolved

**Comment by shawntabrizi** [comment]
@guzzit do you know how to make a companion PR for Polkadot and Cumulus?

**Comment by guzzit** [comment]
Hi @shawntabrizi, no I don't 😅

**Comment by guzzit** [comment]
Hi @shawntabrizi, can I copy the remove_approval weight function from substrate to polkadot's pallet_treasury::WeightInfo struct for the companion pr?

**Comment by guzzit** [comment]
Hi @shawntabrizi, I added the polkadot companion pr: [paritytech/polkadot#5208](https://github.com/paritytech/polkadot/pull/5208)

**Pull request closed without merging**